cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -445,133 +447,182 @@ func (m *McmManager) SetMachineDeploymentSize(machinedeployment *MachineDeployme
 }
 
 // DeleteMachines deletes the Machines and also reduces the desired replicas of the Machinedeplyoment in parallel.
-func (m *McmManager) DeleteMachines(machines []*Ref) error {
+func (m *McmManager) DeleteMachines(machinesAskedForRemoval []*Ref) error {
 	var (
-		mdclone                             *v1alpha1.MachineDeployment
 		terminatingMachines                 []*v1alpha1.Machine
 		expectedToTerminateMachineNodePairs []machineNameNodeNamePair
 	)
-
-	if len(machines) == 0 {
+	if len(machinesAskedForRemoval) == 0 {
 		return nil
 	}
-	commonMachineDeployment, err := m.GetMachineDeploymentForMachine(machines[0])
+	commonMachineDeployment, err := m.GetMachineDeploymentForMachine(machinesAskedForRemoval[0])
+	if err != nil {
+		return err
+	}
+	if err = m.canDeleteMachines(commonMachineDeployment.Name, machinesAskedForRemoval); err != nil {
+		return err
+	}
+	// update priority of all machines in the machineDeployment to 3 except the ones that have nodes with the ToBeDeleted taint
+	if err = m.resetPriorityOnMachinesWithNodesWithoutToBeDeletedTaintInMachineDeployment(commonMachineDeployment.Name); err != nil {
+		return err
+	}
+	// update priorities of machines to be deleted except the ones already in termination to 1
+	for _, machine := range machinesAskedForRemoval {
+		// Trying to update the priority of machine till retryDeadline
+		retryDeadline := time.Now().Add(maxRetryTimeout)
+		for {
+			mc, err := m.getMachineUntilDeadline(machine.Name, conflictRetryInterval, retryDeadline)
+			if err != nil {
+				return err
+			}
+			mclone := mc.DeepCopy()
+			if isMachineTerminating(mclone) {
+				terminatingMachines = append(terminatingMachines, mclone)
+				break
+			} else {
+				expectedToTerminateMachineNodePairs = append(expectedToTerminateMachineNodePairs, machineNameNodeNamePair{mclone.Name, mclone.Labels["node"]})
+			}
+			if err = m.updateAnnotationOnMachineUntilDeadline(mclone.Name, machinePriorityAnnotation, "1", conflictRetryInterval, retryDeadline); err != nil {
+				return err
+			}
+			// Break out of loop when update succeeds
+			klog.Infof("Machine %s of machineDeployment %s marked with priority 1 successfully", machine.Name, commonMachineDeployment.Name)
+			break
+		}
+	}
+	// Trying to update the machineDeployment till the deadline
+	updatedReplicas, err := m.scaleDownMachineDeploymentUntilDeadline(commonMachineDeployment.Name, int32(len(machinesAskedForRemoval)-len(terminatingMachines)), conflictRetryInterval, time.Now().Add(maxRetryTimeout))
```

Comment:
was it not possible to just take the length of expectedToTerminateMachineNodePairs instead bothering to get machines which are already in terminating state and then subtracting from len(machinesAskedForRemoval)

Reply:
addressed in [03c13fc](https://github.com/gardener/autoscaler/pull/215/commits/03c13fc3b37527d0e3ca8a8571237abdb64ec68f)